evm-validate/src/lib.rs

```diff
@@ -417,3 +422,331 @@ fn validate_transaction_hashes(
+
+    // iterate over the block numbers and receipts roots
+    for (block_num, block_receipts_root) in block_numbers.iter().zip(block_receipts_root.iter()) {
+        let block_num = block_num.unwrap();
+        let receipts_root = block_receipts_root.unwrap().try_into().unwrap();
+        expected_receipts_root_by_block_num_mapping.insert(block_num, receipts_root);
+    }
+    
+    for (block_num, expected) in expected_receipts_root_by_block_num_mapping.iter() {
+        let calculated = receipts_root_by_block_num_mapping.get(block_num).unwrap();
+        if expected != calculated {
+            return Err(anyhow!("Receipts root mismatch at block {}.\nExpected:\n{},\nFound:\n{:?}", block_num, expected, calculated));
+        }
+    }
+
+    Ok(())
+}
+
+fn extract_log_cols_as_arrays(logs: &RecordBatch) -> Result<(
+        &PrimitiveArray<UInt64Type>,
+        &PrimitiveArray<UInt64Type>,
+        &GenericByteArray<GenericBinaryType<i32>>,
+        &GenericByteArray<GenericBinaryType<i32>>,
+        &GenericByteArray<GenericBinaryType<i32>>,
+        &GenericByteArray<GenericBinaryType<i32>>,
+        &GenericByteArray<GenericBinaryType<i32>>,
+        &GenericByteArray<GenericBinaryType<i32>>
+    )> {
+    let log_block_nums = logs
+        .column_by_name("block_number")
+        .context("get log block num col")?
+        .as_any()
+        .downcast_ref::<UInt64Array>()
+        .context("get log block num col as u64")?;
+
+    let log_tx_idx = logs
+        .column_by_name("transaction_index")
+        .context("get tx index column")?
+        .as_any()
+        .downcast_ref::<UInt64Array>()
+        .context("get tx index col as u64")?;
+    
+    let log_address = logs
+        .column_by_name("address")
+        .context("get address column")?
+        .as_any()
+        .downcast_ref::<BinaryArray>()
+        .context("get address as binary")?;
+
+    let log_topic0 = logs
+        .column_by_name("topic0")
+        .context("get topic0 column")?
+        .as_any()
+        .downcast_ref::<BinaryArray>()
+        .context("get topic0 as binary")?;
+
+    let log_topic1 = logs
+        .column_by_name("topic1")
+        .context("get topic1 column")?
+        .as_any()
+        .downcast_ref::<BinaryArray>()
+        .context("get topic1 as binary")?;
+
+    let log_topic2 = logs
+        .column_by_name("topic2")
+        .context("get topic2 column")?
+        .as_any()
+        .downcast_ref::<BinaryArray>()
+        .context("get topic2 as binary")?;
+
+    let log_topic3 = logs
+        .column_by_name("topic3")
+        .context("get topic3 column")?
+        .as_any()
+        .downcast_ref::<BinaryArray>()
+        .context("get topic3 as binary")?;
+    
+    let log_data = logs
+        .column_by_name("data")
+        .context("get data column")?
+        .as_any()
+        .downcast_ref::<BinaryArray>()
+        .context("get data as binary")?;
```

Comment:
Raise errors if values have NULL, when they shouldn't.

Reply:
This can be a seperate field validation function which validates fields aren't null or zero when they shouldn't be